##################################################################
### Configuration for SNPS assessment and dataset optimization ###
##################################################################

# set path to relevant folders and files and set contig type


# set path to the HybPiper base folder that contains the assembly results (e.g. "~/HybPiper_assembly/" )
path_to_hybpiper_results = ""

# set path for HybPhaser output files (e.g. "~/HybPhaser_supercontig/" ). This folder will contain most files generated by HybPhaser. 
path_for_HybPhaser_output = ""

# direct to the file with target sequences (baits) (e.g., "/path_to_target_capture_sequencing/targets.fasta" )
fasta_file_with_targets = ""

# set path to the text file with names of accessions (one line per accession) (e.g. "~/HybPiper_assembly/nameslist.txt")
txt_file_with_list_of_accessions = ""

# set contig type for standard HybPiper assembly ("normal") or for the supercontig assembly ("supercontig") when using intronerate.py to generate concatenated exon sequenes
contig = "normal"          # "normal" or "supercontig"



### Dataset optimization 
#########################

# set variables to define thresholds for the dataset optimization and run the script from the main script

### Missing data 

# set threshold for each locus how much missing data (samples not sequenced) is required to be removed
# recommend setting to "1.5*IQR", which removes outliers
threshold_exclude_loci_x_missing_samples = "none"  # e.g. "none", "1.5*IQR", 10, 100, "proportion"
proportion_of_missing_samples_per_locus = 0.1 # only applies when threshold_exclude_loci_x_missing_samples = "proportion"

# set threshold for each sample how much missing data (loci not sequenced) is required to be removed
# recommend setting at first to "none" in order to not exclude any samples, but can be adjusted to remove pooor samples
threshold_exclude_samples_x_missing_loci = "none"  # e.g. "none", "1.5*IQR", 100, 200, "proportion"
proportion_of_missing_loci_per_sample = 0.1 # only applies when threshold_exclude_samples_x_missing_loci = "proportion"

# set threshold for minimum recovered sequence length for a sample to be included (as proportion of the target sequence length)
# any sample with less than the given proportion of the target sequence length is removed from the dataset
threshold_exclude_samples_x_proportion_of_target_sequence = "none"


### Paralogs 

# set threshold of mean % of SNPs across all samples for a locus to be removed 
threshold_exclude_loci_paralogs_4all_propSNP = "1.5*IQR" # e.g. "none", "1.5*IQR", 0.01, 0.05

# set whether outlier loci are to be removed per sample
remove_paralogs_4each = "yes"       # "yes" or "no"


### Sequence list generation

# The fasta files for the normal contig sequences (not supercontigs) in HybPiper do not contain the gene names. 
# When generating seuquence lists for samples, gene sequences do not contain the gene name, but only the sample name. 
# This can be corrected, but the step is time consuming and can be skipped with setting the next variable to "no".

write_gene_names_in_contig_sample_seqlist = "no"  # "yes" or "no"
