##################################################################
### Configuration for SNPS assessment and dataset optimization ###
##################################################################

# General settings
path_to_hybpiper_results = ""
path_for_HybPhaser_output = ""
fasta_file_with_targets = ""
targets_file_format = "DNA"    # "DNA" or "AA" 
txt_file_with_list_of_accessions = ""
contig = "normal"    # "normal" or "supercontig"

# Dataset optimization 
remove_samples_with_less_than_this_propotion_of_loci_recovered = 0
remove_samples_with_less_than_this_propotion_of_target_sequence_length_recovered = 0
remove_loci_with_less_than_this_propotion_of_samples_recovered = 0
remove_loci_with_less_than_this_propotion_of_target_sequence_length_recovered = 0

# Paralogs 
remove_loci_for_all_samples_with_more_than_this_mean_proportion_of_SNPs = "outliers"   # any number between 0 and 1 or "outliers" 
file_with_putative_paralogs_to_remove_for_all_samples = ""
remove_outlier_loci_for_each_sample = "yes"       

# Generating sequence lists
write_gene_names_in_contig_sample_seqlist = "no"


# Description of variables
##########################

### General settings

# path_to_hybpiper_results 
## set path to the HybPiper base folder that contains the assembly results (e.g. "~/HybPiper_assembly/" )

# path_for_HybPhaser_output
## set path for HybPhaser output files (e.g. "~/HybPhaser_supercontig/" ). This folder will contain most files generated by HybPhaser. 

# fasta_file_with_targets
## direct to the file with target sequences (baits) (e.g., "/path_to_target_capture_sequencing/targets.fasta" )

# targets_file_format
## set targets file format: "DNA" for nucleotides or "AA" for amino acids

# txt_file_with_list_of_accession
## set path to the text file with names of accessions (one line per accession) (e.g. "~/HybPiper_assembly/nameslist.txt")

# contig
## set contig type for standard HybPiper assembly ("normal") or for the supercontig assembly ("supercontig") when using intronerate.py to generate concatenated exon sequenes


### Missing data 

# remove_samples_with_less_than_this_propotion_of_loci_recovered
## set threshold for samples to be removed that have a low proportion of the number of loci recovered
## e.g. 0.8 will only keep samples that have a sequence recovered for at least 80% of loci

# remove_samples_with_less_than_this_propotion_of_target_sequence_length_recovered 
## set threshold for samples to be removed that have short sequence length recovered (as proportion of the target sequence length) 
## e.g. 0.6 will only keep samples that have on average 60% of the target sequence length recovered

# remove_loci_with_less_than_this_propotion_of_samples_recovered
## set threshold for loci to be removed that have a low proportion of the number of samples recovered
## e.g. 0.8 will only keep loci that have a sequence recovered for at least 80% of samples

# remove_loci_with_less_than_this_propotion_of_target_sequence_length_recovered
## set threshold for loci to be removed that have short sequence length recovered (as proportion of the target sequence length) 
## e.g. 0.6 will only keep loci that have on average 60% of the target sequence length recovered


### Paralogs 

#remove_loci_for_all_samples_with_more_than_this_mean_proportion_of_SNPs
## any number between 0 and 1 or "outliers" 
## set threshold to remove loci with high mean proportions of SNPs across all samples.
## e.g. 0.02 will remove all loci with an average of 2% or SNPS across all samples
## "outliers" will remove all loci that have more than 1.5*IQR (interquartile range) above the 3rd quartile of mean SNPs. 
## "file" will remove all loci specified in a text file

#file_with_putative_paralogs_to_remove_for_all_samples
## or alternatively remove all loci based on a list of loci in a text file (one line per locus)
## this can be used for the processing a subset of accessions and the list of paralogs from another analysis should be imported  (e.g. of phased accessions)
## OPTIONAL! Using "" will result in applying previous selected mode or threshold

# remove_outlier_loci_for_each_sample
## set whether outlier loci are to be removed per sample
## "yes" will remove for each samples loci that have a higher proportion of SNPs than 1.5*IQR (interquartile range) above the 3rd quartile


### Generating sequence lists

# write_gene_names_in_contig_sample_seqlist = "no"
## HybPhaser collects sequences from the HybPiper folders to generate sequence lists using the unix cat command for each locus and also each sample.
## Due to the lack of gene names in the normal contig sequences (not the supercontigs) this is not possible 
## for the generation of sequence lists for each samples of the normal denovo contigs. 
## This is circumvented by a solution using find, generating temporary files, renaming the sequence names in each file and then collating them. 
## If you do not need the gene names in the sequence lists for denovo contigs per sample and want to save time, change the next variable to"no"!



