##################################################################
### Configuration for SNPS assessment and dataset optimization ###
##################################################################


### General settings
#####################

# set path to the HybPiper base folder that contains the assembly results (e.g. "~/HybPiper_assembly/" )
path_to_hybpiper_results = ""

# set path for HybPhaser output files (e.g. "~/HybPhaser_supercontig/" ). This folder will contain most files generated by HybPhaser. 
path_for_HybPhaser_output = ""

# direct to the file with target sequences (baits) (e.g., "/path_to_target_capture_sequencing/targets.fasta" )
fasta_file_with_targets = ""

# set path to the text file with names of accessions (one line per accession) (e.g. "~/HybPiper_assembly/nameslist.txt")
txt_file_with_list_of_accessions = ""

# set contig type for standard HybPiper assembly ("normal") or for the supercontig assembly ("supercontig") when using intronerate.py to generate concatenated exon sequenes
contig = "normal"          # "normal" or "supercontig"



### Dataset optimization 
#########################

# You can leave these settings as they are and check the graphs and tables first before deciding on thresholds that make sense for your dataset

### Missing data 
# remove samples and loci from the dataset that have on average low amount of sequences and/or short sequences recovered

# set threshold for samples to be removed that have a low proportion of the number of loci recovered
# e.g. 0.8 will only keep samples that have a sequence recovered for at least 80% of loci
remove_samples_with_less_than_this_propotion_of_loci_recovered = 0

# set threshold for samples to be removed that have short sequence length recovered (as proportion of the target sequence length) 
# e.g. 0.6 will only keep samples that have on average 60% of the target sequence length recovered
remove_samples_with_less_than_this_propotion_of_target_sequence_length_recovered = 0


# set threshold for loci to be removed that have a low proportion of the number of samples recovered
# e.g. 0.8 will only keep loci that have a sequence recovered for at least 80% of samples
remove_loci_with_less_than_this_propotion_of_samples_recovered = 0

# set threshold for loci to be removed that have short sequence length recovered (as proportion of the target sequence length) 
# e.g. 0.6 will only keep loci that have on average 60% of the target sequence length recovered
remove_loci_with_less_than_this_propotion_of_target_sequence_length_recovered = 0



### Paralogs 

# set threshold to remove loci with high mean proportions of SNPs across all samples.
# e.g. 0.02 will remove all loci with an average of 2% or SNPS across all samples
# "outliers" will remove all loci that have more than 1.5*IQR (interquartile range) above the 3rd quartile of mean SNPs. 
remove_loci_for_all_samples_with_more_than_this_mean_proportion_of_SNPs = "outliers"   # any number between 0 and 1 or "outliers" 

# set whether outlier loci are to be removed per sample
# "yes" will remove for each samples loci that have a higher proportion of SNPs than 1.5*IQR (interquartile range) above the 3rd quartile
remove_outlier_loci_for_each_sample = "yes"       # "yes" or "no"



### Generating sequence lists

# HybPhaser collects sequences from the HybPiper folders to generate sequence lists using the unix cat command for each locus and also each sample.
# Due to the lack of gene names in the normal contig sequences (not the supercontigs) this is not possible 
# for the generation of sequence lists for each samples of the normal denovo contigs. 
# This is circumvented by a solution using find, generating temporary files, renaming the sequence names in each file and then collating them. 
#
# If you do not need the gene names in the sequence lists for denovo contigs per sample and want to save time, change the next variable to"no"!

write_gene_names_in_contig_sample_seqlist = "yes"


