#!/bin/bash
############
set -e
set -u
set -o pipefail

NAMESLIST="nameslist.txt"
BASE_FOLDER="./"
OUTPUT_FOLDER=$BASE_FOLDER/'_mapped_reads'

while getopts 'n:p:o:' OPTION; do
  case "$OPTION" in

    n)
       NAMESLIST=$OPTARG
     ;;
    
    p)
      BASE_FOLDER=$OPTARG
     ;;         
    
    o)
      OUTPUT_FOLDER=$OPTARG
     ;;         
    
    ?)
      echo "script usage: $(basename $0) [-n file with samples names (nameslist.txt)] [-b HybPiper base folder] [-o output folder]"
      echo "This script is used to extract reads that mapped sucessfully to the target sequences generating a new read file for the clade association analysis. The mapped reads are extracted from the BAM file generated by HybPiper. 
      
      Usage: extract_mapped_reads.sh [options]
      
        Options:
      
            -n  Text file with list of samples (one line per samples)
        
            -p  Path to HybPiper results folder. Default is current folder.
        
            -o  Path to output folder where new read files are saved. Default is a folder called _mapped_reads in the HybPiper results folder."
     
      exit 1
      ;;
  esac
done
shift "$(($OPTIND -1))"

mkdir -p $OUTPUT_FOLDER/

while read NAME;
do
if [ $NAME != "" ]; then 
    if [ -f $BASE_FOLDER/$NAME/$NAME".bam" ]; then        
        echo Extracting mapped reads from $NAME
        #samtools fastq -F 4 $BASE_FOLDER/$NAME/$NAME".bam" > $OUTPUT_FOLDER/$NAME".fastq"
        samtools view -F 4 -b $BASE_FOLDER/$NAME/$NAME".bam"| samtools sort -n -o $OUTPUT_FOLDER/$NAME"_sort.bam"
        samtools fixmate -m $OUTPUT_FOLDER/$NAME"_sort.bam" $OUTPUT_FOLDER/$NAME"_fixm.bam"
        samtools sort -o $OUTPUT_FOLDER/$NAME"_fixmsort.bam" $OUTPUT_FOLDER/$NAME"_fixm.bam"
        samtools markdup -r -s $OUTPUT_FOLDER/$NAME"_fixmsort.bam" $OUTPUT_FOLDER/$NAME"_dupm.bam"
        samtools fastq $OUTPUT_FOLDER/$NAME"_dupm.bam" > $OUTPUT_FOLDER/$NAME".fastq"
        rm $OUTPUT_FOLDER/$NAME"_sort.bam" $OUTPUT_FOLDER/$NAME"_fixm.bam" $OUTPUT_FOLDER/$NAME"_fixmsort.bam" $OUTPUT_FOLDER/$NAME"_dupm.bam"
    else 
        echo Skipped extracting reads for $NAME. No BAM file found.
    fi
fi
done < $NAMESLIST




