#!/bin/bash
############
set +e
set -u
set -o pipefail

NAMESLIST="nameslist.txt"
BASE_FOLDER="./"
OUTPUT_FOLDER=$BASE_FOLDER/'_mapped_reads'
READ_TYPE="paired"

while getopts 'n:p:o:s' OPTION; do
  case "$OPTION" in

    n)
       NAMESLIST=$OPTARG
     ;;
    
    p)
      BASE_FOLDER=$OPTARG
     ;;         
    
    o)
      OUTPUT_FOLDER=$OPTARG
     ;;         
    
    s)
      READ_TYPE="single"
     ;;


    ?)
      echo "script usage: $(basename $0) [-n file with samples names (nameslist.txt)] [-b HybPiper base folder] [-o output folder]"
      echo "This script is used to extract reads that mapped sucessfully to the target sequences generating a new read file for the clade association analysis. The mapped reads are extracted from the BAM file generated by HybPiper. 
      
      Usage: extract_mapped_reads.sh [options]
      
        Options:
      
            -n  Text file with list of samples (one line per samples)
        
            -p  Path to HybPiper results folder. Default is current folder.
        
            -o  Path to output folder where new read files are saved. Default is a folder called _mapped_reads in the HybPiper results folder.

            -s use when reads are single-end reads"
     
      exit 1
      ;;
  esac
done
shift "$(($OPTIND -1))"

mkdir -p $OUTPUT_FOLDER/

if [ $READ_TYPE == "single" ]; then
	while read NAME;
	do
	if [ $NAME != "" ]; then 
		echo "collecting mapped reads for: $NAME"
		if ! [ -z "$BASE_FOLDER/$NAME/*/*_unpaired.fasta" ]; then
			cat $BASE_FOLDER/$NAME/*/*"_unpaired.fasta" > $OUTPUT_FOLDER/$NAME".fasta"
		else 
			echo "no unpaired files for $NAME"
		fi
	fi    
	done < $NAMESLIST
else
	while read NAME;
	do
	if [ $NAME != "" ]; then
		echo "collecting mapped reads for: $NAME"	
		if ! [ -z "$BASE_FOLDER/$NAME/*/*_interleaved.fastq" ]; then
			cat $BASE_FOLDER/$NAME/*/*"_interleaved.fastq" \
			| paste - - - - - - - - \
			| tee >(cut -f 1-4 | tr "\t" "\n" > $OUTPUT_FOLDER/$NAME"_R1.fastq") \
			| cut -f 5-8 | tr "\t" "\n" > $OUTPUT_FOLDER/$NAME"_R2.fastq"
		elif ! [ -z "$BASE_FOLDER/$NAME/*/*_interleaved.fasta" ]; then
			cat $BASE_FOLDER/$NAME/*/*"_interleaved.fasta" \
			| paste - - - - - - - - \
			| tee >(cut -f 1-2 | tr "\t" "\n" > $OUTPUT_FOLDER/$NAME"_R1.fasta") \
			| cut -f 3-4 | tr "\t" "\n" > $OUTPUT_FOLDER/$NAME"_R2.fasta"
		else 
			echo "no interleaved files for $NAME"
		fi    
	fi    
	done < $NAMESLIST
fi







